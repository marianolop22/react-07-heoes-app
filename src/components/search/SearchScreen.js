import React, { useMemo } from 'react'
import { useLocation } from 'react-router-dom';
import { HeroCard } from '../heros/HeroCard';
import { useForm } from '../../hooks/useForm';
import queryString from 'query-string';
import { getHerosByName } from '../../selectors/getHerosByName';

export const SearchScreen = ({history}) => {

    const location = useLocation();
    const {q=''} = queryString.parse(location.search);
    const [{searchText}, handleInputChange] = useForm ( {searchText:q} );
    
    const herosFiltered = useMemo(()=>getHerosByName(q), [q]) ;
    
    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    };

    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr/>
                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            placeholder="Find your Hero"
                            className="form-control"
                            name="searchText"
                            value={searchText}
                            onChange={handleInputChange}
                            autoComplete="off"
                        />
                        <button
                            type="submit"
                            className="btn mt-1 btn-block btn-primary"
                        >
                            Search...
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>

                    {
                        (q==='') && 
                            <div className="alert alert-info">
                                Search a hero
                            </div>
                    }
                    {
                        (q!=='' && herosFiltered.length === 0 ) && 
                            <div className="alert alert-danger">
                                No heros with {q}
                            </div>
                    }
                    

                    {
                        herosFiltered.map( hero => {
                            return (
                                <HeroCard className="pb-2"
                                    key={hero.superhero}
                                    {...hero}
                                />
                            )
                        })
                    }
                </div>

            </div>
        </div>
    )
}
