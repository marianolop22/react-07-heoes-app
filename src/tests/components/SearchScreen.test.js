import '@testing-library/jest-dom';
import React from 'react';
import { mount } from 'enzyme';
import { SearchScreen } from '../../components/search/SearchScreen';
import { MemoryRouter, Route, Router } from 'react-router-dom';
import { AuthContext } from '../../auth/AuthContext';

describe ('Prueba de SearchScreen.test', ()=> {

    const historyMock = {
        push: jest.fn(),
    }

    const contextValue = {
        user: {name:'Mariano', logged:true},
        dispatch: jest.fn()
    };

    const wrapper = mount (
        <MemoryRouter initialEntries={['/search']}>
            <Route 
                path="/search"
                component={ (props) => <SearchScreen history={historyMock} />}
            />
        </MemoryRouter>
    );

    test('debe de renderizar el componente', () => {
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
    });

    test('debe de mostrar a batman y y el input con el valor del query sting', () => {
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route 
                    path="/search"
                    component={ (props) => <SearchScreen history={historyMock} />}
                />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('HeroCard').exists()).toBe(true);
    })
    
    test('debe de mostrar un error con hero que no existe', () => {
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search?q=123']}>
                <Route 
                    path="/search"
                    component={ (props) => <SearchScreen history={historyMock} />}
                />
            </MemoryRouter>
        );

        expect( wrapper.find('.alert-danger').text().trim() ).toBe(`No heros with 123`);
    })


    test('debe de llamar el push del history', () => {
        const wrapper = mount (
            <MemoryRouter initialEntries={['/search']}>
                <Route 
                    path="/search"
                    component={ (props) => <SearchScreen history={historyMock} />}
                />
            </MemoryRouter>
        );

        //esto es para cambiar la caja de texto de un componente
        wrapper.find('input').simulate('change', {
            target: {
                name: 'searchText', //nombre del input
                value: 'batman!!!'
            }
        })

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(wrapper).toMatchSnapshot();
        expect(historyMock.push).toHaveBeenCalledWith(`?q=batman!!!`);




    })







});