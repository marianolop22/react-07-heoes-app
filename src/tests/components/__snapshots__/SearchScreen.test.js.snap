// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prueba de SearchScreen.test debe de llamar el push del history 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value="batman!!!"
        />
        <button
          className="btn mt-1 btn-block btn-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`Prueba de SearchScreen.test debe de mostrar a batman y y el input con el valor del query sting 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-1 btn-block btn-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heros/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small>
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Mas...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Prueba de SearchScreen.test debe de renderizar el componente 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your Hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-1 btn-block btn-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
